### ISSUE TITLE
Preregistration #2: Terminal-Bench 2.1 development baseline for Stella 0.6.21 (#909, supersedes #950)

### ISSUE THREAD

**Issue by macanderson** [macanderson/stella#1002]
Preregistration for the **second and definitive** attempt at #909's measured Terminal-Bench 2.1 run. Filed before any paid call, including the sentinel. Not edited after data.

The first attempt is **[abandoned and published in full](https://github.com/macanderson/stella/issues/950#issuecomment-5136230666)** — all four of its trials, including the zeros. Nothing from it carries into this run.

## Why there is a second attempt

The first run measured a SUT with two defects **in its own measurement path**, both of which have since been fixed and merged:

| | |
|---|---|
| [#960](https://github.com/macanderson/stella/issues/960) → [#970](https://github.com/macanderson/stella/pull/970) | `stella run` completed its turn and never exited. Every trial burned its entire agent timeout and recorded `AgentTimeoutError` **even when it scored 1.0**, so wall clock was ~3x and the exception telemetry carried no signal. |
| [#973](https://github.com/macanderson/stella/issues/973) → [#981](https://github.com/macanderson/stella/pull/981) | The verification ladder was blind on all three evidence channels at once and reported a confident *"the file likely does not exist"* about a file that was in the container. It now abstains when blind and counts what the recorder counted. |

The second one is the reason a re-run matters rather than merely being faster. #909 exists to test the claim that *the verification ladder produces better outcomes*. Measuring a build whose ladder cannot see the work would have produced a number that could not speak to that claim in either direction.

Both fixes are verified present in this SUT, by commit:

```
9e23ad8  fix(run): `stella run` exits when the turn ends (#960) (#970)
87e3f32  fix(verify): abstain when the ladder is blind, and count what the recorder counted (#973, #927) (#981)
```

## System under test

| Field | Value |
|---|---|
| SUT commit | `250976d3fd3b64e5db94aded4b1e6e08a63dca17` |
| Relationship to `main` | **unmodified `origin/main`** at freeze — no local patch |
| `git describe` | `v0.5.68-124-g250976d` (version `0.6.21`) |
| Binary SHA-256 | `b55053cae2f82387d132a5eedc673d6abb25db1c2ffee421cc33f71a8c8fc55c` |
| Build | native `x86_64-unknown-linux-gnu`, `cargo build --release --locked` |
| Build stamp | `STELLA_BUILD_GIT_SHA=250976d3fd3b64e5db94aded4b1e6e08a63dca17` |

`--locked` resolves cleanly here only because [#992](https://github.com/macanderson/stella/pull/992) landed first: the 0.6.18 bump had left `stella-engine` and `stella-runtime` at 0.6.17 in `Cargo.lock`, which broke `cargo install --locked` — the documented install route — and was invisible from any warm checkout. It surfaced on the first clean clone, which is this host.

## Host — and what changed about it

| Field | Value |
|---|---|
| Instance | AWS `m6id.2xlarge`, us-east-1, **native x86_64** |
| CPU / memory | 8 vCPU / 30 GiB, dedicated |
| Docker | 29.6.2, 89 task images pre-pulled with retries, 0 failures |

Two host artifacts present in the abandoned run are **gone** here, and both had visibly distorted it:

* **Rosetta emulation** is absent. The first run measured within ~5% of native, which was defensible but not the same as absent.
* **The 0.6 MB/s link is gone.** `torch-tensor-parallelism` previously returned **no verdict at all** because its verifier could not finish downloading 2.8 GiB of CUDA wheels inside its 1800s ceiling — a zero in the score with no agent involvement. The same image now pulls in 1.3 seconds.

This is still **not** the audited claim host: no committed host attestation, no Management-API-minted spend-capped key, no intent ledger, no `secure_launcher`, no external maintainer review. It remains a development baseline.

## Benchmark and configuration

| Field | Value |
|---|---|
| Dataset | `terminal-bench/terminal-bench-2-1@sha256:7d7bdc1cbedad549fc1140404bd4dc45e5fd0ea7c4186773687d177ad3a0699a` |
| Tasks | 89 — the whole set, no filters, no exclusions |
| Harbor | `0.6.1` (an audited constant, not an upgrade candidate) |
| Model | `openrouter/z-ai/glm-5.1` — the protocol's mandatory primary and the public Claude Code comparator's model |
| Attempts | 1 per task · Retries 0 · Concurrency **3**, single phase |
| Per-trial budget cap | `STELLA_BUDGET=0.60` USD |
| Reflection | `STELLA_DISABLE_REFLECTION=1` |
| Engine posture | adapter canonical: `default_model` routing only, auto/effort/reasoning off, `headless_scope_bypass: on`, per-role effort high with triage low |

**Single phase, not the A/B split.** The split existed only because the previous host had less memory than the largest task. It was always a property of the host and never of the measurement; this host fits the largest task beside its neighbours, so all 89 run in one job. Task set, attempt count and denominator are unchanged.

`headless_scope_bypass: on` remains **score-affecting** and is disclosed as the protocol requires: with it off, a headless trial with no operator to approve an over-threshold plan self-terminates any plan over 5 steps.

## Analysis plan — fixed before any data

Unchanged from the first preregistration, because changing it now would be the whole problem.

**Primary metric.** pass@1 = (tasks whose trial has external-verifier reward exactly `1.0`) / **89**.

**The denominator is fixed at 89.** A trial that errors, times out, crashes, hits the budget cap, or fails for any operational reason scores 0 and stays in. A task producing no row also scores 0. Nothing is dropped or re-run — including failures that are obviously infrastructure's fault.

**Reward semantics** are taken from the in-repo claim analyzer, not reinvented: the external verifier's reward is authoritative, and a trial that timed out having already left a correct solution keeps the reward it earned.

**Interval.** 95% CI both ways: nonparametric bootstrap over the 89 per-task outcomes (50,000 resamples, seed `20260729`, percentile), and Clopper–Pearson exact as a seed-free cross-check.

> Note on precision, stated in advance so the result is not oversold: at 89 tasks the 95% interval is roughly **±20 percentage points** whatever the score. More attempts per task would buy pass@k semantics and lower per-task noise — **not** a narrower interval, because the bootstrap is over tasks. This run therefore supports "Stella scores approximately X" and does **not** support "Stella beats Y by N points" for any small N.

**Secondary, descriptive, no thresholds.** Total USD; prompt/completion tokens; per-task wall clock; trials that hit the budget cap; trials that hit the agent timeout; trials with zero tool calls; **trials with no verifier verdict** (the ceiling on how much of the score is infrastructure rather than agent); trials Stella reported `completed`.

**Stop conditions.** Operational failure only — balance, Docker, credentials, host. Any stop is disclosed and the partial published as partial. No outcome-triggered stop: a bad early score is not a reason to stop, change anything, or start over.

**Sentinel gate.** One paid synthetic trial must return external-verifier reward exactly `1.0` before the 89 are launched. It is excluded from all analysis.

## Budget

OpenRouter balance at filing: **$74.37**. Expected spend ~$15–25 (89 trials at the $0.17–0.28/trial the protocol's estimate and the abandoned run's observed mean bracket). Ceiling **$40**, measured against Harbor's own per-trial cost records rather than the account balance, because the key is shared with unrelated work. AWS ~$0.48/hr, expected $3–6.

## Reproduction

`bench/evidence/run/` is committed and parameterised; `cloud_bootstrap.sh` provisions a host and runs this end to end. The evidence lands in `bench/evidence/<run-id>/` with the manifest, `trials.jsonl`, the per-task table, and the score — recomputable from `trials.jsonl` alone.

**Comment by macanderson** [comment]
## Disclosure: the benchmark posture disables the witness tier

Filed at 5 of 89 trials, with **zero information about the final score**, because a caveat published after the number is worth much less than one published before it.

### What the trials show

`#981` is working as intended. Where no channel can see the turn, the ladder now abstains instead of asserting a false negative:

```
verification could not be performed: UNVERIFIABLE — no evidence channel could
observe this turn, so nothing is claimed about it (this is NOT a finding ...)
```

and where evidence exists it is counted: *"while 3 file change events occurred…"*. That is the #973 fix doing its job.

### What it does not fix, and cannot

```
continuing without an authored witness: no author independent of the worker
(judge and worker both resolved to `openrouter/z-ai/glm-5.1`)
```

The witness requires an author independent of the worker. The adapter's canonical benchmark posture pins `default_model` with `allowed_models: ["openrouter/z-ai/glm-5.1"]` for **every** role — worker, judge, triage alike. Under that posture the witness can never be authored, on any task, in any run.

This is not a defect in `#981`. It is a property of the posture, and it was the third of the three failures listed in #973 — the only one that is a configuration decision rather than a code bug.

### Why it is being disclosed rather than fixed mid-run

Changing the posture now would change the system under test after trials had been observed, which the fixed-denominator and no-rerun rules exist to forbid. The posture is frozen for this run. It is stated here so the number is read correctly.

### How to read the resulting number

**This run measures Stella with the witness tier structurally off.** The remaining tiers — triage, planning, the judge, deterministic checks — are active, and the judge's evidence channels now work or abstain honestly.

That matters specifically for the claim #909 was opened to test. "The verification ladder produces better outcomes" cannot be fully evaluated by a run in which one tier of that ladder is disabled by the harness. Whatever the score, it is a **lower bound** on what the full ladder does, not a measurement of it.

### What would measure the full ladder

Pinning a second model for the judge/witness role — the protocol's roster already names DeepSeek V4 Pro and Grok-4.5 alongside GLM-5.1 — would let the witness be authored. That is a **separate, separately-preregistered run**, and comparing it against this one would be the actual experiment: same tasks, same SUT, witness off versus witness on. This run is the control arm of that comparison whether or not the treatment arm is ever run.

**Comment by macanderson** [comment]
# Result: 38/89 = 42.70%

The first measured Terminal-Bench result Stella has ever had.

```
pass@1                  38/89 = 42.70%
95% CI bootstrap        [32.58%, 52.81%]   seed 20260729, 50,000 draws
95% CI Clopper-Pearson  [32.26%, 53.63%]
spend                   $23.05
```

SUT `250976d3` (unmodified `origin/main`, ancestor of `main`), binary `b55053ca…`, Harbor 0.6.1, dataset pinned by digest, posture hashed. Evidence in `bench/evidence/tb21-20260730c/`; the score recomputes from `trials.jsonl` alone, verified on a second machine from a fresh checkout.

## Against the comparator

**Claude Code + GLM-5.1 = 58.7%** on the public 2.1 leaderboard — same model, same 89 tasks, same verifier.

58.7% is **outside** this run's upper bound (53.6%). On this evidence Stella is behind Claude Code at the same model by roughly 16 points. Stating it plainly because the preregistration exists to stop it being stated any other way.

## Qualifications, both preregistered, both biasing downward

**15 of 89 trials hit the $0.60 per-trial budget cap** — truncated mid-work, not beaten. At 17% of the run the cap is a live term in the score, not a rare guard. Raising it is the single change most likely to move the number.

**The witness tier was structurally off** ([#1007](https://github.com/macanderson/stella/issues/1007)): the adapter posture pins one model for every role, and the witness requires an author independent of the worker. This is a lower bound on the full ladder.

## Reading the descriptive fields

| field | value | |
|---|---:|---|
| trials with an exception | 46 | **not** a failure rate — see below |
| trials Stella reported completed | 74 | |
| trials with no verifier verdict | 2 | the ceiling on infrastructure contamination |
| trials that hit the budget cap | 15 | truncated, biasing down |
| trials with zero tool calls | 7 | |

46 exceptions against 74 completions: the exception count is dominated by harness bookkeeping, not by failure. Those two fields sit side by side in `score.json` for exactly this reason.

Infrastructure contamination was **2 of 89**, against the abandoned laptop attempt where the single heavy task that finished returned no verdict at all.

## Tasks passed

`adaptive-rejection-sampler`, `break-filter-js-from-html`, `build-pmars`, `cobol-modernization`, `compile-compcert`, `configure-git-webserver`, `constraints-scheduling`, `crack-7z-hash`, `custom-memory-heap-crash`, `feal-differential-cryptanalysis`, `feal-linear-cryptanalysis`, `financial-document-processor`, `fix-code-vulnerability`, `fix-git`, `fix-ocaml-gc`, `git-leak-recovery`, `git-multibranch`, `hf-model-inference`, `log-summary-date-ranges`, `mcmc-sampling-stan`, `merge-diff-arc-agi-task`, `modernize-scientific-stack`, `multi-source-data-merger`, `nginx-request-logging`, `polyglot-c-py`, `polyglot-rust-c`, `portfolio-optimization`, `prove-plus-comm`, `pytorch-model-cli`, `query-optimize`, `reshard-c4-data`, `rstan-to-pystan`, `sam-cell-seg`, `schemelike-metacircular-eval`, `sparql-university`, `sqlite-db-truncate`, `sqlite-with-gcov`, `vulnerable-secret`

## Tasks failed

`bn-fit-modify`, `build-cython-ext`, `build-pov-ray`, `caffe-cifar-10`, `cancel-async-tasks`, `chess-best-move`, `circuit-fibsqrt`, `code-from-image`, `count-dataset-tokens`, `db-wal-recovery`, `distribution-search`, `dna-assembly`, `dna-insert`, `extract-elf`, `extract-moves-from-video`, `filter-js-from-html`, `gcode-to-text`, `gpt2-codegolf`, `headless-terminal`, `install-windows-3.11`, `kv-store-grpc`, `large-scale-text-editing`, `largest-eigenval`, `llm-inference-batching-scheduler`, `mailman`, `make-doom-for-mips`, `make-mips-interpreter`, `model-extraction-relu-logits`, `mteb-leaderboard`, `mteb-retrieve`, `openssl-selfsigned-cert`, `overfull-hbox`, `password-recovery`, `path-tracing`, `path-tracing-reverse`, `protein-assembly`, `pypi-server`, `pytorch-model-recovery`, `qemu-alpine-ssh`, `qemu-startup`, `raman-fitting`, `regex-chess`, `regex-log`, `sanitize-git-repo`, `torch-pipeline-parallelism`, `torch-tensor-parallelism`, `train-fasttext`, `tune-mjcf`, `video-processing`, `winning-avg-corewars`, `write-compressor`

## What this does and does not license

It supports: *"Stella scores about 43% on Terminal-Bench 2.1 with GLM-5.1, behind Claude Code's 58.7% at the same model."*

It does not support any claim of the form *"Stella beats X by N points"* for small N — at 89 tasks the interval is ±~10pp whatever the spend, because the bootstrap is over tasks. That was stated in the preregistration before the data, and it still holds now that the data is unflattering.

It is **not** a leaderboard row: no host attestation, no Management-API-minted key, no intent ledger, no `secure_launcher`, no external maintainer review. `claim_eligibility.audited_public_row` is `false` in the manifest.

**Comment by macanderson** [comment]
## RETRACTED — this run is not a valid test, and no number from it should be quoted

The result comment above reports 38/89 = 42.70% against a comparator at 58.65%. **Withdraw that comparison entirely.** The measurement is not comparable to the baseline it was measured against.

### Three constraints applied to Stella and not to the comparator

| Constraint | Comparator | This run |
|---|---|---|
| Effort | `max` (protocol §Comparator: "GLM-5.1 at max effort") | **`high`** |
| Witness tier | full agent | **disabled** — posture pins one model for every role, witness needs an independent author |
| Per-trial budget | none | **$0.60 hard cap** — bound on 15 of 89 trials, truncated 12 mid-work |

Each biases the result downward. Together they mean the run measures a deliberately constrained configuration and compares it to an unconstrained one.

### Also: conclusions drawn from unverified quantities

After the score landed I compared token totals and cost per trial and drew efficiency conclusions. Those do not stand either:

- Stella's 88.5M "tokens" are **94.4% cache reads** (82.1M of 87.0M input). The comparator's 398,783,761 is a published figure whose accounting I never saw. Putting those two numbers in a ratio and reporting "1.11× the tokens" to three significant figures was precision standing in for rigour.
- The cost difference was read as an efficiency win. The protocol states plainly that this is a **different-provider-route** comparison, so route pricing is at least as likely an explanation as anything about the agent.

### What the protocol already said

> The existing public Claude result is not contemporaneous. It is suitable for a transparent public-leaderboard product comparison, but not for claiming that all difference is caused by the CLI. **A later matched study must run both agents contemporaneously with the same model endpoint and account tier.**

That is the study that should have been designed. It was not.

### What survives

Only this, and it is not a comparison: **a Stella build scored 38/89 on Terminal-Bench 2.1 under a $0.60/trial cap, at effort `high`, with the witness tier off.** The raw per-trial data is preserved for reference. PR #1012 is closed and will not be merged.

### The replacement

A matched study, with the posture defects fixed first (#1007):

1. Effort `max` for default/worker/judge, matching the comparator.
2. No arbitrary budget cap — or one derived from the comparator's actual per-trial spend rather than from protecting an account balance.
3. Witness enabled via a second pinned author model.
4. **Both agents run contemporaneously** — Stella and Claude Code, same host, same 89 tasks, same GLM-5.1 endpoint, same account tier, same constraints. Harbor supports `-a claude-code` directly, so this is a configuration difference, not new infrastructure.

Point 4 is what removes the remaining unverifiable assumptions about token accounting and provider routing, because both sides are then measured by the same instrument on the same day.
